README.markdown
# Render Debugging Toolbar for Rails

If you are looking at the front end of a Rails app and see something you want to
work on, there are several ways you might answer the question:

## Which template is rendering that particular chunk of HTML?

If you have simple views, logical markup and your URLs map very quickly to
views, then you can probably just open your editor and guess pretty much
exactly where you need to edit. You're in the Rails happy place, and you
probably don't need a tool to help you.

If you're working on a less-happy Rails application, however, you might have
some of the following impediments:

 * hundreds of similarly-named views
 * complex views with partials within partials
 * URLs that don't map so easily quickly and obviously to view filenames
 * markup without obviously unique patterns to search for

I was in this situation recently; I found myself wishing for something
to just tell me where the HTML was coming from.

# The Render Debugging Toolbar helps you find your view

There are tools that let you interactively point at the different parts of
your page and tell you a bunch of back-end information.

Drupal has the [Drupal Theme Developer][1], Django has the
[Django Debug Toolbar][2], but I couldn't find anything to
help me out when I asked Rails the same question of this Rails app:
"What is generating that HTML right there?"

The Render Debugging Toolbar aims to be that tool for Rails.

# Installation

  Add the following to the "development" group in your Gemfile:
    gem "rails-render-debugging-toolbar"
  Bundler will take care of the rest when you run
    bundle install
  
  If you're not using Bundler, you won't have a Gemfile, so you'll need
  to install the gem manually.

# Usage

The toolbar needs a lot of extra markup to do its work, so you probably won't
want to leave it enabled all the time. You can turn it on for a single request
by adding the query parameter "?debug=render" to a request.

For example, if you are interested in the page:

    http://localhost:3000/examples/new

You can try the toolbar by visiting the following:

    http://localhost:3000/examples/new?template_debug

When the toolbar is enabled, you should see a little debugging checkbox overlaid
on top of the normal page. Switch it on, and then everything you point at should
show a panel of debugging information about what rendered it.

Simply point at the element you're interested in, and the panel will tell you
where the view is.


# Dependencies
The Rails Debugging Toolbar works with Rails 2.3.5 for sure, and it should
work with Rails 3.0.0 as well.

The toolbar also needs jQuery to do its job, and will load jQuery automatically
if it's not already on the page. One day I would like to make the toolbar work
with no external dependencies but right now I'm happy with jQuery.


[1]: http://drupal.org/project/devel_themer
[2]: https://github.com/robhudson/django-debug-toolbar


# License

Copyright 2011 Rob Hunter and ThoughtWorks, Inc

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


History.txt
== 0.0.3 (2011-02-14)
=== Bugfixes
* The rendering output should stay inside HTML <body> tags, so
  as not to trigger quirks mode rendering or any other weirdness.
* jQuery is loaded automatically in case it's not already present,
  rather than just failing.

=== New features
* Works with Rails 3.0 as well as Rails 2.3
* Better variable inspection -- hashes should show up in table-like things.
* You can turn off debug-follows-pointer and scroll around the detail



== 0.0.2 (2011-01-31)

Initial gem release.



lib/rails-debugging-toolbar.rb
require 'ext/action_controller'
module RailsDebuggingToolbar
  module Extensions
    module ActionView
      def render(options = {}, local_assigns = {}, &block)
        actual_output = super(options, local_assigns, &block)
        id = next_available_render_id

        on_entering_render
        record_render_details(id, options, local_assigns)
        on_leaving_render

        return debug_log_after_body(actual_output) if at_outer_level_render?

        wrapped_output(actual_output, id)
      end

      private
      attr_writer :my_render_depth
      def render_partial(options = {})
        actual_output = super(options)
        id = next_available_render_id

        on_entering_render
        record_render_details(id, options, {})
        on_leaving_render

        wrapped_output(actual_output, id)
      end

      def on_entering_render
        self.my_render_depth = my_render_depth.succ
      end
      def on_leaving_render
        self.my_render_depth = my_render_depth.pred
      end

      def wrapped_output(actual_output, id)
        ERB.new(<<-HTML).result(binding)
          <span class='render-debug partial' id='render-debug-wrapper-<%= h id %>'>
            <%= actual_output %>
          </span>
        HTML
      end

      def debug_log_after_body(actual_output)
        debug_log = ERB.new(<<-HTML).result(binding)
          <div class='render-debug' id='debug-log'>
            <% recorded_render_details.each_pair do |id, stuff| %>
              <%
              recorded_options = stuff[:options]
              locals = recorded_options[:locals] || {}
              partial = recorded_options[:partial] || "unknown"
              unrecognized_options = recorded_options.reject do |option_name, option_value|
                [:locals, :partial].include? option_name
              end
              %>
              <div class="render-debug render-detail" id="render-debug-detail-<%= h id %>">
                <h3><label for="render-debug-wrapper-<%= h id %>"><code><%= h partial %></code></label></h3>
                <% if locals.any? %>
                  <h4><label for="render-debug-locals-<%= h id %>">Locals</label></h4>
                  <dl>
                    <% locals.each_pair do |local_name, local_value| %>
                      <dt><%= h local_name %></dt>
                      <dd><code><%= h local_value.inspect %></code></dd>
                    <% end %>
                  </dl>
                <% end %>

                <% if unrecognized_options.any? %>
                  <h4><label for="render-debug-options-<%= h id %>">Other options</label></h4>
                  <dl>
                    <% unrecognized_options.each_pair do |option_name, option_value| %>
                      <dt><%= h option_name %></dt>
                      <dd><%= h option_value.inspect %></dd>
                    <% end %>
                  </dl>
                <% end %>
              </div>
            <% end %>
          </div>
          <form action="#" class="render-debug debug-show">
            <input type="checkbox" id="enable-debug-detail-checkbox" name="render" value="debug" />
            <label for="enable-debug-detail-checkbox">Show rendering details</label>
            <input type="checkbox" id="debug-follows-cursor-checkbox" name="follow_cursor" value="yes" />
            <label for="debug-follows-cursor-checkbox">Show rendering details</label>
          </form>
          <style type="text/css">
            #debug-log {
              display: block; position: absolute; top: 0px; right: 0px; top: 0px; width: 300px; z-index: -1000;
              background: transparent;
              font-family: sans-serif;
              text-align: left;
              color: #ccc;
              border: none;
            }
            #debug-log.active {
              display: block; position: fixed; top: 0px; right: 0px; top: 0px; width: 300px; z-index: 1000;
              background: black; opacity: 0.8;
              border: thin solid white;
            }
            #debug-log .render-detail {
              display: none;
            }
            #debug-log .render-detail.active {
              display: block;
              border-top: thin dashed #777;
              padding-bottom: 1ex;
              padding-top: 1ex;
            }
            #debug-log h1,
            #debug-log h2,
            #debug-log h3,
            #debug-log h4,
            #debug-log h5,
            #debug-log h6
            {
              font-weight: bold;
              display: block;
              padding-top: 1ex;
              font-size: 20px;
              color: white;
            }
            #debug-log h3
            {
              color: #f13;
              text-align: center;
              height: auto;
              width: 80%;
              padding-left: 1em;
              padding-right: 1em;
            }
            #debug-log code {
              white-space: pre;
            }
            #debug-log dl {
              display: block;
            }
            #debug-log dt {
              display: block;
              margin-left: 0px;
              font-weight: bold;
              float: left;
              width: 8em;
              word-break: break-all;
              overflow: hidden;
            }
            #debug-log dd {
              display: block;
              margin-left: 2px;
              max-height: 4ex;
              overflow: hidden;
            }
            form.debug-show {
              display: block; position: fixed; bottom: 0px; left: 0px; width: 100px; z-index: 1000;
              background: black; opacity: 1.0;
              border: thin solid white;
            }
          </style>
          
          <script type="text/javascript">
          (function ($) {
            $(function() {
              var checkbox = $("input#enable-debug-detail-checkbox");
              
              checkbox.change(function() {
                if ($(this).is(':checked')) {
                  $("#debug-log").addClass("active");
                  $(".render-debug.partial").hover(function() {
                    var detail_div = $("label[for=" + this.id + "]").parents("#debug-log .render-detail");
                    detail_div.addClass("active");
                  }, function() {
                    var detail_div = $("label[for=" + this.id + "]").parents("#debug-log .render-detail");
                    detail_div.removeClass("active");
                  });
                } else {
                  $("#debug-log").removeClass("active");
                  $(".render-debug.partial").unbind();
                }
              });
            });
          })(jQuery);
          </script>
        HTML
        actual_output.sub("</body>", debug_log + "</body>".html_safe!)
      end

      def my_render_depth
        @some_render_depth ||= 0
        @some_render_depth.tap {|x| puts "Render depth is #{x}"}
      end

      def at_outer_level_render?
        (my_render_depth == 0)
      end

      def next_available_render_id
        @render_id_counter ||= 0
        @render_id_counter += 1 # XXX: totally not thread safe
        @render_id_counter
      end

      def record_render_details(id, options, local_assigns)
        recorded_render_details[id] = {:options => options, :local_assigns => local_assigns}
      end

      def recorded_render_details
        @recorded_render_details ||= {}
      end
    end
  end
end


lib/rails_debugging_toolbar/extensions.rb
require 'sha1'
module RailsDebuggingToolbar
  module Extensions
    module ActionController
      def view_context
        returning(super) do |view|
          extension_module = ::RailsDebuggingToolbar::Extensions::ActionView
          view.extend(extension_module) unless view.kind_of? extension_module
        end
      end
    end
    module ActionView
      def render(options = {}, local_assigns = {}, &block)
        actual_output = super(options, local_assigns, &block)
        id = next_available_render_id

        record_render_details(id, options, local_assigns)

        return debug_log_after_body(actual_output) if actual_output.include? "</body>"

        wrapped_output(actual_output, id)
      end

      private
      attr_writer :my_render_depth
      def render_partial(options = {})
        actual_output = super(options)
        id = next_available_render_id

        on_entering_render
        record_render_details(id, options, {})
        on_leaving_render

        wrapped_output(actual_output, id)
      end

      def on_entering_render
        self.my_render_depth = my_render_depth.succ
      end
      def on_leaving_render
        self.my_render_depth = my_render_depth.pred
      end

      def wrapped_output(actual_output, id)
        open_wrapping = raw("<span class='render-debug partial' id='render-debug-wrapper-#{ h id}'>")
        close_wrapping = raw("</span>")
        
        # insert the wrapping spans, but staying inside any <body> / </body>
        output_with_open_wrapping = actual_output.rpartition(/<body\b.*?>/).insert(2, open_wrapping).join
        output_with_both_wrapping = output_with_open_wrapping.partition('</body>').insert(1, close_wrapping).join
        
        # open_wrapping + actual_output + close_wrapping
        raw(output_with_both_wrapping)
      end

      def debug_log_after_body(actual_output)
        debug_log = ERB.new(<<-HTML).result(binding)
          <div class='render-debug' id='debug-log'>
            <% recorded_render_details.each_pair do |id, stuff| %>
              <%
              recorded_options = stuff[:options]
              locals = recorded_options[:locals] || {}
              partial = recorded_options[:partial] || "unknown"
              unrecognized_options = recorded_options.reject do |option_name, option_value|
                [:locals, :partial].include? option_name
              end
              %>
              <div class="render-debug render-detail" id="render-debug-detail-<%= h id %>">
                <h3><label for="render-debug-wrapper-<%= h id %>"><code><%= h partial %></code></label></h3>
                <% if locals.any? %>
                  <h4><label for="render-debug-locals-<%= h id %>">Locals</label></h4>
                  <%= print_hash_as_html_for_debugging(locals) %>
                <% end %>

                <% if unrecognized_options.any? %>
                  <h4><label for="render-debug-options-<%= h id %>">Other options</label></h4>
                  <%= print_hash_as_html_for_debugging(unrecognized_options) %>
                <% end %>
              </div>
            <% end %>
          </div>
          <form action="#" class="render-debug" id="debug-show">
            <input type="checkbox" id="enable-debug-detail-checkbox" name="render" value="debug" accesskey="r" />
            <label for="enable-debug-detail-checkbox">Show rendering details (<kbd>r</kbd)>)</label>
            <input type="checkbox" id="debug-follows-cursor-checkbox" name="follow_cursor" checked="checked" accesskey="f" />
            <label for="debug-follows-cursor-checkbox">Debug follows cursor (<kbd>f</kbd)>)</label>
          </form>
          <style type="text/css">
            #debug-log {
              display: block; position: absolute; top: 0px; right: 0px; top: 0px; width: 300px; z-index: -1000;
              background: transparent;
              font-family: sans-serif;
              text-align: left;
              color: #ccc;
              border: none;
              overflow: hidden;
              height: 4px; /* just enough to know it's there */
            }
            #debug-log.active {
              display: block; position: fixed; top: 0px; right: 0px; top: 0px; width: 300px; z-index: 1000;
              background: rgba(0,0,0,0.8);
              color: #ccc;
              border: thin solid white;
              overflow: auto;
              height: 100%;
            }
            #debug-log .render-detail {
              display: none;
            }
            #debug-log .render-detail.active {
              display: block;
              border-top: thin dashed #777;
              padding-bottom: 1ex;
              padding-top: 1ex;
            }
            #debug-log h1,
            #debug-log h2,
            #debug-log h3,
            #debug-log h4,
            #debug-log h5,
            #debug-log h6
            {
              font-weight: bold;
              display: block;
              padding-top: 1ex;
              font-size: 20px;
              color: white;
            }
            #debug-log h3
            {
              color: #f13;
              text-align: center;
              height: auto;
              width: 80%;
              padding-left: 1em;
              padding-right: 1em;
            }
            #debug-log code {
              white-space: pre;
            }
            #debug-log dl {
              display: block;
            }
            #debug-log dt {
              display: block;
              margin-left: 0px;
              font-weight: bold;
            }
            #debug-log dd {
              display: block;
              margin-left: 2px;
              max-height: 4ex;
            }
            form#debug-show {
              display: block; position: fixed; bottom: 0px; left: 0px; width: 30em; z-index: 1000;
              background: rgba(0,0,0,1.0);
              color: #ccc;
              border: thin solid white;
            }
          </style>          
          <script type="text/javascript">//<![CDATA
          // Load JQuery if it's not already present
          if (typeof(jQuery) === 'undefined') {
            document.write('<scr' + 'ipt type="text/javascript" src="//ajax.googleapis.com/ajax/libs/jquery/1.4.4/jquery.min.js"></scr' + 'ipt>');
          }
          // ]>
          </script>
          <script type="text/javascript">//<![CDATA
          (function ($) {
            $(function() {
              var show_bar_checkbox = $("input#enable-debug-detail-checkbox");
              var follow_cursor_checkbox = $("input#debug-follows-cursor-checkbox");
              var should_follow_cursor = $(follow_cursor_checkbox).is(':checked');
              
              show_bar_checkbox.change(function() {
                if ($(show_bar_checkbox).is(':checked')) {
                  $("#debug-log").addClass("active");
                } else {
                  $("#debug-log").removeClass("active");
                  $(".render-debug.partial").unbind();
                }
              });
              follow_cursor_checkbox.change(function() {
                should_follow_cursor = $(follow_cursor_checkbox).is(':checked');
                if (should_follow_cursor) {
                  // a clean slate for mousing around
                  $("#debug-log .render-detail.active").removeClass("active");
                }
              });
              $(".render-debug.partial").hover(function() {
                if (!should_follow_cursor) { return true; }
                
                var detail_div = $("label[for=" + this.id + "]").parents("#debug-log .render-detail");
                detail_div.addClass("active");
              }, function() {
                if (!should_follow_cursor) { return true; }
                
                var detail_div = $("label[for=" + this.id + "]").parents("#debug-log .render-detail");
                detail_div.removeClass("active");
              });
              
            });
          })(jQuery);
          //]>
          </script>
        HTML
        output_with_debug_log = actual_output.sub("</body>", debug_log + raw("</body>"))
        @debug_log_rendered = true
        output_with_debug_log
      end

      def my_render_depth
        @some_render_depth ||= 0
      end

      def at_outer_level_render?
        (my_render_depth == 0)
      end

      def next_available_render_id
        @render_id_counter ||= 0
        @render_id_counter += 1 # XXX: totally not thread safe
        @render_id_counter
      end

      def record_render_details(id, options, local_assigns)
        recorded_render_details[id] = {:options => options, :local_assigns => local_assigns}
      end

      def recorded_render_details
        @recorded_render_details ||= {}
      end
      
      def print_as_html_for_debugging(subject)
        return h subject.relative_path if subject.respond_to? :relative_path
        return h subject.path if subject.respond_to? :path
        return print_as_html_for_debugging if subject.respond_to? :each_pair
        ERB.new("<code><%= h subject.inspect %></code>").result(binding)
      end
      
      def print_hash_as_html_for_debugging(hsh)
        ERB.new(<<-HTML).result(binding)
          <dl>
            <% hsh.each_pair do |key, value| %>
              <dt><%= h key %></dt>
              <dd><%= print_as_html_for_debugging(value) %></code></dd>
            <% end %>
          </dl>
        HTML
      end

    end
  end
end
